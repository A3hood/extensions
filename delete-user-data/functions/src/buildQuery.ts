import { DocumentData, Query, WhereFilterOp } from "@google-cloud/firestore";
import * as admin from "firebase-admin";

export const buildQuery = async (userId): Promise<Query<DocumentData>[]> => {
  const db = admin.firestore();
  const promises = [];
  const documents = [];

  const docs = await admin
    .firestore()
    .collection("queries")
    .listDocuments();

  for await (const doc of docs) {
    const document = await doc.get();

    const data = (documents[doc.id] = document.data());

    const ref = db.collectionGroup(data.collection === "{uid}" ? `${userId}` : data.collection);
    const condition = (data.conditions || [])[0];

    if (!condition || !condition.length) {
      promises.push(
        new Promise((resolve) => {
          ref.get().then((querySnapshot) => {
            querySnapshot.forEach(async (doc) => {
              console.log(doc.id, " => ", doc.data());
              await data.recusrive ? db.recursiveDelete(doc.ref): db.doc(doc.ref.path).delete();
            });
          });
          resolve(true);
        })
      );
    }
    

    if (condition?.where) {
      promises.push(
        new Promise((resolve) => {
          ref
            .where(
              condition.where[0] as string,
              condition.where[1] as WhereFilterOp,
              condition.where[2].replace("{uid}", userId) as string
            )
            .get()
            .then((querySnapshot) => {
              querySnapshot.forEach(async (doc) => {
                console.log(doc.id, " => ", doc.data());
                await data.recusrive ? db.recursiveDelete(doc.ref): db.doc(doc.ref.path).delete();        
              });
              resolve(true);
            });
        })
      );
    }
  }

  console.log("heres >>>", promises);

  const result = await Promise.all(promises);

  console.log("result >>>", result);

  return promises;
};
