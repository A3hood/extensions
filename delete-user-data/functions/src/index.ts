/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as admin from "firebase-admin";
import * as functions from "firebase-functions";
import * as firebase_tools from "firebase-tools";
import { getEventarc } from "firebase-admin/eventarc";

import { getDatabaseUrl } from "./helpers";

import config from "./config";
import * as logs from "./logs";

// Helper function for selecting correct domain adrress
const databaseURL = getDatabaseUrl(
  config.selectedDatabaseInstance,
  config.selectedDatabaseLocation
);

// Initialize the Firebase Admin SDK
admin.initializeApp({
  credential: admin.credential.applicationDefault(),
  databaseURL,
});

const eventChannel =
  process.env.EVENTARC_CHANNEL &&
  getEventarc().channel(process.env.EVENTARC_CHANNEL, {
    allowedEventTypes: process.env.EXT_SELECTED_EVENTS,
  });

logs.init();

/*
 * The clearData function removes personal data from the RealTime Database,
 * Storage, and Firestore. It waits for all deletions to complete, and then
 * returns a success message.
 */
export const clearData = functions.auth.user().onDelete(async (user) => {
  logs.start();

  const { firestorePaths, rtdbPaths, storagePaths, queryCollection } = config;
  const { uid } = user;

  /** Add query record deletions here */
  if (queryCollection) {
    // const queries = await buildQueries();
  }

  if (eventChannel) {
    const paths = firestorePaths.split(",").map((path) => {
      return path.replace("{UID}", uid);
    });

    await eventChannel.publish({
      type: "firebase.extensions.delete-user-data.v1.trigger",
      subject: "test",
      data: {
        paths,
      },
    });
  }
});

const clearDatabaseData = async (databasePaths: string, uid: string) => {
  logs.rtdbDeleting();

  const paths = extractUserPaths(databasePaths, uid);
  const promises = paths.map(async (path) => {
    try {
      logs.rtdbPathDeleting(path);
      await admin
        .database()
        .ref(path)
        .remove();
      logs.rtdbPathDeleted(path);
    } catch (err) {
      logs.rtdbPathError(path, err);
    }
  });

  await Promise.all(promises);

  logs.rtdbDeleted();
};

const clearStorageData = async (storagePaths: string, uid: string) => {
  logs.storageDeleting();

  const paths = extractUserPaths(storagePaths, uid);
  const promises = paths.map(async (path) => {
    const parts = path.split("/");
    const bucketName = parts[0];
    const bucket =
      bucketName === "{DEFAULT}"
        ? admin.storage().bucket(config.storageBucketDefault)
        : admin.storage().bucket(bucketName);
    const prefix = parts.slice(1).join("/");
    try {
      logs.storagePathDeleting(prefix);
      await bucket.deleteFiles({
        prefix,
      });
      logs.storagePathDeleted(prefix);
    } catch (err) {
      if (err.code === 404) {
        logs.storagePath404(prefix);
      } else {
        logs.storagePathError(prefix, err);
      }
    }
  });

  await Promise.all(promises);

  logs.storageDeleted();
};

const clearFirestoreData = async (firestorePaths: string, uid: string) => {
  logs.firestoreDeleting();

  const paths = extractUserPaths(firestorePaths, uid);
  const promises = paths.map(async (path) => {
    try {
      const isRecursive = config.firestoreDeleteMode === "recursive";

      if (!isRecursive) {
        const firestore = admin.firestore();
        logs.firestorePathDeleting(path, false);

        // Wrapping in transaction to allow for automatic retries (#48)
        await firestore.runTransaction((transaction) => {
          transaction.delete(firestore.doc(path));
          return Promise.resolve();
        });
        logs.firestorePathDeleted(path, false);
      } else {
        logs.firestorePathDeleting(path, true);
        await firebase_tools.firestore.delete(path, {
          project: process.env.PROJECT_ID,
          recursive: true,
          yes: true, // auto-confirmation
        });
        logs.firestorePathDeleted(path, true);
      }
    } catch (err) {
      logs.firestorePathError(path, err);
    }
  });

  await Promise.all(promises);

  logs.firestoreDeleted();
};

const extractUserPaths = (paths: string, uid: string) => {
  return paths.split(",").map((path) => replaceUID(path, uid));
};

const replaceUID = (path: string, uid: string) => {
  return path.replace(/{UID}/g, uid);
};
