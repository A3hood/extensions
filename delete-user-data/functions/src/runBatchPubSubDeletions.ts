const { PubSub } = require("@google-cloud/pubsub");
var _ = require("lodash");

import * as config from "./config";

export async function runBatchPubSubDeletions(paths: string[]) {
  /** Define pubsub */
  const pubsub = new PubSub();

  if (!paths.length) return Promise.resolve();

  /** Define batch array variables */
  for await (const chunk of _.chunk(paths, 450)) {
    const topic = pubsub.topic(
      `projects/${process.env.GOOGLE_CLOUD_PROJECT ||
        process.env.PROJECT_ID}/topics/${config.default.deletionTopic}`
    );
    await topic.publish(Buffer.from(JSON.stringify({ paths: chunk })));
  }
}
