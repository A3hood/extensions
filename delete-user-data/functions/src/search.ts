import * as admin from "firebase-admin";
import * as config from "./config";
const { PubSub } = require("@google-cloud/pubsub");

export const search = async (
  uid,
  document?: admin.firestore.DocumentReference<
    admin.firestore.DocumentData
  > | null
) => {
  const db = admin.firestore();

  const pubsub = new PubSub();

  const topic = pubsub.topic(
    `projects/${process.env.GOOGLE_CLOUD_PROJECT ||
      process.env.PROJECT_ID}/topics/${config.default.searchTopic}`
  );

  const collections = !document
    ? await db.listCollections()
    : await document.listCollections();

  for (const collection of collections) {
    topic.publish(Buffer.from(JSON.stringify({ path: collection.path, uid })));
  }
};
