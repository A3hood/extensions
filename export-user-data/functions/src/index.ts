/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as admin from "firebase-admin";
import * as functions from "firebase-functions";
import * as csvStringify from "csv-stringify/sync";


const STORAGE_BUCKET = 'storage-bucket';
const STORAGE_EXPORT_DIRECTORY = "exports";

// Initialize the Firebase Admin SDK
admin.initializeApp({
  credential: admin.credential.applicationDefault(),
});

const getPaths = (uid: string) => {
  const configParam = "users/{UID},posts/{UID}";

  const paths = configParam
    .split(",")
    .map((path) => path.replace(/{UID}/g, uid));

  const collections: string[] = [];
  const docs: string[] = [];

  for (let path of paths) {
    const parts = path.split("/");
    if (parts.length % 2 === 0) {
      docs.push(path);
    } else {
      collections.push(path);
    }
  }

  return {
    collections,
    docs,
  };
};

const uploadToStorage = async (
  csv: string,
  uid: string,
  exportId: string,
  path: string
) => {
  const formattedPath = path.replace(/\//g, "_");

  const storagePath = `${STORAGE_EXPORT_DIRECTORY}/${uid}/${exportId}/${formattedPath}.csv`;

  const file = admin
    .storage()
    .bucket(STORAGE_BUCKET)
    .file(storagePath);

  await file.save(csv);

  await admin.firestore()
    .doc(`exports/${exportId}`)
    .update({
      status: "complete",
      storagePath: storagePath,
    })

  return storagePath
};

export const exportUserData = functions.https.onCall(async (_data, context) => {
  const startedAt = admin.firestore.Timestamp.now();
  // const uid = context.auth.uid;
  const uid = "123";

  // create /exports/<exportId> with status "pending"
  const ref = await admin
    .firestore()
    .collection("exports")
    .add({ status: "pending", started_at: startedAt, uid });

  const exportId = ref.id;

  // get all paths we will export
  const { collections, docs } = getPaths(uid);

  const headers = ["TYPE", "path", "data"];

  const promises = [];

  for (let collection of collections) {

    const snap = await admin
      .firestore()
      .collection(collection)
      .get();

    if (!snap.empty) {
      const csvData = snap.docs.map((doc => {

        const path = `${collection}/${doc.id}`;
        return ["FIRESTORE", path, JSON.stringify(doc.data())];
      }));

      csvData.unshift(headers);

      const csv = csvStringify.stringify(csvData);
      promises.push(uploadToStorage(csv, uid, exportId, collection));
    }
  }

  for (let doc of docs) {
    const csvData = [headers];
    const snap = await admin
      .firestore()
      .doc(doc)
      .get();
    if (snap.exists) {
      const data = snap.data();
      csvData.push(["FIRESTORE", doc, JSON.stringify(data)]);

      const csv = csvStringify.stringify(csvData);
      promises.push(uploadToStorage(csv, uid, exportId, doc));
    }
  }

  await Promise.all(promises);

  await admin
    .firestore()
    .collection("exports")
    .doc(exportId)
    .update({
      uid,
      started_at: startedAt,
      status: "complete",
      location: `gs://storage-bucket/exports/${uid}/${exportId}`,
      // TODO include links to export
    });

  // TODO return the storage refs
  return { exportId };
});
