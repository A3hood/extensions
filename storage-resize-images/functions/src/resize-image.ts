import * as os from "os";
import * as sharp from "sharp";
import * as path from "path";
import * as fs from "fs";

import { Bucket, File } from "@google-cloud/storage";
import { ObjectMetadata } from "firebase-functions/lib/providers/storage";
import { uuid } from "uuidv4";

import config, { deleteImage } from "./config";
import * as logs from "./logs";

export interface ResizedImageResult {
  size: string;
  success: boolean;
}

export function resize(originalFile, resizedFile, size) {
  let height, width;
  if (size.indexOf(",") !== -1) {
    [width, height] = size.split(",");
  } else if (size.indexOf("x") !== -1) {
    [width, height] = size.split("x");
  } else {
    throw new Error("height and width are not delimited by a ',' or a 'x'");
  }

  return sharp(originalFile)
    .rotate()
    .resize(parseInt(width, 10), parseInt(height, 10), {
      fit: "inside",
      withoutEnlargement: true,
    })
    .toFile(resizedFile);
}

export const resizeImage = async ({
  bucket,
  originalFile,
  fileDir,
  fileNameWithoutExtension,
  fileExtension,
  contentType,
  size,
  objectMetadata,
  remoteFile,
  filePath,
}: {
  bucket: Bucket;
  originalFile: string;
  fileDir: string;
  fileNameWithoutExtension: string;
  fileExtension: string;
  contentType: string;
  size: string;
  objectMetadata: ObjectMetadata;
  remoteFile: File;
  filePath: string;
}): Promise<ResizedImageResult> => {
  const resizedFileName = `${fileNameWithoutExtension}_${size}${fileExtension}`;
  // Path where resized image will be uploaded to in Storage.
  const resizedFilePath = path.normalize(
    config.resizedImagesPath
      ? path.join(fileDir, config.resizedImagesPath, resizedFileName)
      : path.join(fileDir, resizedFileName)
  );
  let resizedFile: string;

  try {
    resizedFile = path.join(os.tmpdir(), resizedFileName);

    // Cloud Storage files.
    const metadata: { [key: string]: any } = {
      contentDisposition: objectMetadata.contentDisposition,
      contentEncoding: objectMetadata.contentEncoding,
      contentLanguage: objectMetadata.contentLanguage,
      contentType: contentType,
      metadata: objectMetadata.metadata || {},
    };
    metadata.metadata.resizedImage = true;
    if (config.cacheControlHeader) {
      metadata.cacheControl = config.cacheControlHeader;
    } else {
      metadata.cacheControl = objectMetadata.cacheControl;
    }

    // If the original image has a download token, add a
    // new token to the image being resized #323
    if (metadata.metadata.firebaseStorageDownloadTokens) {
      metadata.metadata.firebaseStorageDownloadTokens = uuid();
    }

    // Generate a resized image using Sharp.
    logs.imageResizing(resizedFile, size);

    await resize(originalFile, resizedFile, size);

    logs.imageResized(resizedFile);

    // Uploading the resized image.
    logs.imageUploading(resizedFilePath);
    await bucket.upload(resizedFile, {
      destination: resizedFilePath,
      metadata,
    });
    logs.imageUploaded(resizedFilePath);

    if (config.deleteOriginalFile === deleteImage.onSuccess) {
      if (remoteFile) {
        try {
          logs.remoteFileDeleting(filePath);
          await remoteFile.delete();
          logs.remoteFileDeleted(filePath);
        } catch (err) {
          logs.errorDeleting(err);
        }
      }
    }

    return { size, success: true };
  } catch (err) {
    logs.error(err);
    return { size, success: false };
  } finally {
    try {
      // Make sure the local resized file is cleaned up to free up disk space.
      if (resizedFile) {
        logs.tempResizedFileDeleting(resizedFilePath);
        fs.unlinkSync(resizedFile);
        logs.tempResizedFileDeleted(resizedFilePath);
      }
    } catch (err) {
      logs.errorDeleting(err);
    }
  }
};
