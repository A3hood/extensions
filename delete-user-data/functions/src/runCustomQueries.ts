import * as admin from "firebase-admin";
import { DocumentData, Query, WhereFilterOp } from "@google-cloud/firestore";
import { runBatchPubSubDeletions } from "./runBatchPubSubDeletions";

export const runCustomQueries = async (
  userId
): Promise<Query<DocumentData>[]> => {
  const db = admin.firestore();
  const promises = [];
  const documents = [];

  const docs = await admin
    .firestore()
    .collection("queries")
    .listDocuments();

  for await (const doc of docs) {
    const document = await doc.get();

    const data = (documents[doc.id] = document.data());

    const ref = db.collectionGroup(
      data.collection === "{uid}" ? `${userId}` : data.collection
    );

    for await (const condition of data.conditions || []) {
      if (condition && condition.where) {
        ref.where(
          condition.where[0] as string,
          condition.where[1] as WhereFilterOp,
          condition.where[2].replace("{uid}", userId) as string
        );
      }
    }

    const snapshot = await ref.get();

    if (snapshot.docs) {
      await runBatchPubSubDeletions(snapshot.docs.map((doc) => doc.ref.path));
    }
  }

  return promises;
};
