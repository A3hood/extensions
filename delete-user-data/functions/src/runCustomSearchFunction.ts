import * as admin from "firebase-admin";
import fetch from "node-fetch";
import { runBatchPubSubDeletions } from "./runBatchPubSubDeletions";
import config from "./config";

export const runCustomSearchFunction = async (uid: string): Promise<void> => {
  const db = admin.firestore();

  const response = await fetch(config.searchFunction, {
    method: "post",
    body: JSON.stringify({ data: { uid } }),
    headers: { "Content-Type": "application/json" },
  });

  /** Get user resonse **/
  const responseJson = await response.json();

  /** Run pubsub batch */
  await runBatchPubSubDeletions(responseJson);
};
