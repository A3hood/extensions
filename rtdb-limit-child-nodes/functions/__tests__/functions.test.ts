import functionsConfig from "../src/config";
import * as exportedFunctions from "../src";

describe("extension", () => {
  test("functions configuration detected from environment variables", async () => {
    expect(functionsConfig).toMatchSnapshot();
  });

  test("functions are exported", async () => {
    expect(exportedFunctions.rtdblimit).toBeInstanceOf(Function);
  });
});
