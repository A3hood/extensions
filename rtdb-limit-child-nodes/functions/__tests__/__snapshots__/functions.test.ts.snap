// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extension functions configuration detected from environment variables 1`] = `
Object {
  "location": "europe-west2",
  "maxCount": 2,
  "nodePath": "test_path",
}
`;
